pkg/client/block/client.go

```diff
@@ -105,48 +116,72 @@ func (b *blockClient) Close() {
 	b.eventsReplayClient.Close()
 }
 
-// getLatestBlocks listens to the EventsReplayClient's EventsSequence and updates
-// the latestBlock field with the latest block observed.
-func (b *blockClient) getLatestBlocks(ctx context.Context) {
-	latestBlockCh := b.eventsReplayClient.EventsSequence(ctx).Subscribe(ctx).Ch()
-	for block := range latestBlockCh {
-		b.latestBlockMu.Lock()
-		b.latestBlock = block
-		b.latestBlockMu.Unlock()
-	}
+// forEachBlockEvent asynchronously observes block event notifications from the
+// EventsReplayClient's EventsSequence observable & publishes each to latestBlockPublishCh.
+func (b *blockClient) forEachBlockEvent(ctx context.Context, latestBlockPublishCh chan<- client.Block) {
+	channel.ForEach(ctx, b.eventsReplayClient.EventsSequence(ctx),
+		func(ctx context.Context, block client.Block) {
+			latestBlockPublishCh <- block
+		},
+	)
 }
 
-// getInitialBlock fetches the initial block from the chain and sets it as the
-// latest block observed by the blockClient.
-// This is necessary to ensure that the blockClient has the latest block when
-// it is first created.
-// If b.latestBlock is already set by the getLatestBlocks goroutine, then this
-// function skips setting the initial block.
-func (b *blockClient) getInitialBlock(ctx context.Context, client cometBlockClient) error {
-	queryBlockResult, err := client.Block(ctx, nil)
-	if err != nil {
+// getInitialBlock requests the latest block while concurrently waiting for the
+// next block event, publishing whichever occurs first to latestBlockPublishCh.
+func (b *blockClient) getInitialBlock(ctx context.Context, latestBlockPublishCh chan<- client.Block) error {
+	blockQueryResultCh := make(chan client.Block)
+
+	// Query the latest block asynchronously.
+	queryErrCh := b.queryLatestBlock(ctx, blockQueryResultCh)
+
+	// Wait for either the latest block query response, error, or the first block
+	// event to arrive & use whichever occurs first or return an error.
+	//
+	// NB: #latestBlockReplayObs is a proxy for the events sequence observable
+	// because it is guaranteed to be notified on block events in #goForEachBLockEvent().
+	var initialBlock client.Block
+	select {
+	case initialBlock = <-b.latestBlockReplayObs.Subscribe(ctx).Ch():
```

Comment:
This should actually not set `initialBlock` not publish it; in this case, the initial block is already observed.
```suggestion
	case <-b.latestBlockReplayObs.Subscribe(ctx).Ch():
		return nil	
```